tower/src/util/pending.rs

```diff
@@ -0,0 +1,141 @@
+use std::fmt;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_service::Service;
+
+/// Returns a new `Pending` for the given future.
+///
+/// # Example
+/// ```
+/// use tower::{service_fn, Service, ServiceExt};
+/// use tower::util::pending;
+/// use std::convert::Infallible;
+///
+/// # fn main() {
+/// # async {
+/// let mut pending_svc = pending(Box::pin(async {
+///     let svc = service_fn(|_req: ()| async { Ok::<_, Infallible>("ok") });
+///     Ok::<_, Infallible>(svc)
+/// }));
+///
+/// let svc = pending_svc.ready_and().await.unwrap();
+/// let res = svc.call(()).await.unwrap();
+/// # };
+/// # }
+/// ```
+pub fn pending<F, S, R, E>(future: F) -> Pending<F, S>
+where
+    F: Future<Output = Result<S, E>> + Unpin,
+    S: Service<R, Error = E>,
+{
+    Pending {
+        inner: Inner::Future(future),
+    }
+}
+
+/// A type that implements `Service` for a `Future` that produces a `Service`.
+#[derive(Clone)]
+pub struct Pending<F, S> {
+    inner: Inner<F, S>,
+}
+
+impl<F, S> fmt::Debug for Pending<F, S>
+where
+    S: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        #[derive(Debug)]
+        struct Pending;
+
+        let mut f = f.debug_struct("Pending");
+        let f = match &self.inner {
+            Inner::Future(_) => f.field("service", &Pending),
+            Inner::Service(svc) => f.field("service", svc),
+        };
+        f.finish()
+    }
+}
+
+#[derive(Clone)]
+enum Inner<F, S> {
+    Future(F),
+    Service(S),
+}
+
+impl<F, S, R, E> Service<R> for Pending<F, S>
+where
+    F: Future<Output = Result<S, E>> + Unpin,
```

Comment:
The `Unpin` can simply be removed since `poll_ready` is not `Pin`ning `self`.

Reply:
🤯